File: landrop-main/receiver/android/app/src/test/java/com/landrop/receiver/protocol/DiscoveryTest.kt
package com.landrop.receiver.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class DiscoveryTest {

    private fun canonical() = RawDiscoveryInfo(
        service = DISCOVERY_SERVICE_TYPE,
        hostname = "receiver-device.local",
        port = 45821,
        protocolVersion = 1,
        deviceName = "Android Device",
        extraTxtKeys = listOf("lanDrop_version=0.1.0", "capability=apk_delivery"),
    )

    @Test
    fun serviceTypeMatchesSpecification() {
        assertEquals("_lanDrop._tcp.local", DISCOVERY_SERVICE_TYPE)
    }

    @Test
    fun parsesCanonicalDiscoveryInformation() {
        val info = (parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info
        assertEquals("_lanDrop._tcp.local", info.service)
        assertEquals("receiver-device.local", info.hostname)
        assertEquals(45821, info.port)
        assertEquals(1, info.protocolVersion)
        assertEquals("Android Device", info.deviceName)
        assertTrue(isDiscoverySupportedVersion(info))
    }

    @Test
    fun trimsTheDeviceDisplayName() {
        val info = (parseDiscoveryInfo(canonical().copy(deviceName = "  Android Device  ")) as DiscoveryParseResult.Ok).info
        assertEquals("Android Device", info.deviceName)
    }

    @Test
    fun ignoresUnknownExtraTxtKeys() {
        val info = (parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info
        assertEquals("Android Device", info.deviceName)
        assertEquals(45821, info.port)
    }

    @Test
    fun rejectsEveryMalformedCase() {
        val cases = listOf(
            canonical().copy(service = "_other._tcp.local") to DiscoveryProblem.INVALID_SERVICE,
            canonical().copy(hostname = "") to DiscoveryProblem.INVALID_HOSTNAME,
            canonical().copy(hostname = "has space.local") to DiscoveryProblem.INVALID_HOSTNAME,
            canonical().copy(hostname = "a/b") to DiscoveryProblem.INVALID_HOSTNAME,
            canonical().copy(hostname = "a\\b") to DiscoveryProblem.INVALID_HOSTNAME,
            canonical().copy(port = 0) to DiscoveryProblem.INVALID_PORT,
            canonical().copy(port = 65536) to DiscoveryProblem.INVALID_PORT,
            canonical().copy(protocolVersion = 1_500_000_000_000) to DiscoveryProblem.INVALID_PROTOCOL_VERSION,
            canonical().copy(deviceName = "   ") to DiscoveryProblem.INVALID_DEVICE_NAME,
        )
        for ((raw, problem) in cases) {
            assertEquals(problem, (parseDiscoveryInfo(raw) as DiscoveryParseResult.Failing).problem)
        }
    }

    @Test
    fun acceptsUnsupportedVersionAsValidMetadataButFlagsIt() {
        for (version in listOf(-1L, 0L, 2L)) {
            val info = (parseDiscoveryInfo(canonical().copy(protocolVersion = version)) as DiscoveryParseResult.Ok).info
            assertFalse(isDiscoverySupportedVersion(info))
        }
    }

    @Test
    fun registryCollapsesDuplicateAdvertisements() {
        val registry = DiscoveryRegistry()
        val info = (parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info
        registry.upsert(info)
        registry.upsert(info)
        assertEquals(1, registry.size)
        assertEquals("Android Device", registry.list().first().deviceName)
    }

    @Test
    fun identifiesPeersByHostnamePort() {
        assertEquals("receiver-device.local:45821", deviceIdentity("receiver-device.local", 45821))
        val registry = DiscoveryRegistry()
        registry.upsert((parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info)
        registry.upsert((parseDiscoveryInfo(canonical().copy(hostname = "other-device.local")) as DiscoveryParseResult.Ok).info)
        assertEquals(2, registry.size)
    }

    @Test
    fun removesAndClearsDeterministically() {
        val registry = DiscoveryRegistry()
        registry.upsert((parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info)
        registry.remove("receiver-device.local", 45821)
        assertEquals(0, registry.size)
        registry.upsert((parseDiscoveryInfo(canonical()) as DiscoveryParseResult.Ok).info)
        assertEquals(1, registry.size)
        registry.clear()
        assertTrue(registry.list().isEmpty())
    }
}

File: landrop-main/sender/desktop/src/protocol/discovery.test.ts
import { describe, expect, it } from "vitest";

import {
  createDiscoveryRegistry,
  deviceIdentity,
  DISCOVERY_PROBLEMS,
  DISCOVERY_SERVICE_TYPE,
  isDiscoverySupportedVersion,
  parseDiscoveryInfo,
} from "./discovery";
import type { DiscoveryInfo, DiscoveryInfoInput, DiscoveryProblem } from "./discovery";

const fixtures: Record<string, unknown> = require(
  "../../../../protocol/fixtures/model-cases.json",
);

const discovery = fixtures.discovery as {
  service_type: string;
  required_txt_keys: string[];
  peer: DiscoveryInfo;
  peer_with_extra_txt: DiscoveryInfo & { extra_txt_keys: string[] };
  unsupported_version_peer: DiscoveryInfo;
  invalid_discovery_infos: {
    name: string;
    mutations: Record<string, unknown>;
    problem: DiscoveryProblem;
  }[];
  device_identity: { id_format: string };
  events: { event: string; peer: string; note?: string }[];
};

function withMutations(
  base: DiscoveryInfoInput,
  mutations: Record<string, unknown>,
): DiscoveryInfoInput {
  const out: Record<string, unknown> = { ...base };
  for (const [key, value] of Object.entries(mutations)) {
    if (value === null) {
      delete out[key];
    } else {
      out[key] = value;
    }
  }
  return out;
}

describe("phase 05 — discovery foundation", () => {
  it("uses the fixed service type from the protocol specification", () => {
    expect(DISCOVERY_SERVICE_TYPE).toBe("_lanDrop._tcp.local");
    expect(discovery.service_type).toBe(DISCOVERY_SERVICE_TYPE);
    expect(discovery.required_txt_keys).toEqual(["protocol_version", "device_name"]);
  });

  it("parses the canonical discovery information", () => {
    const result = parseDiscoveryInfo(discovery.peer);
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.info).toEqual({
        service: "_lanDrop._tcp.local",
        hostname: "receiver-device.local",
        port: 45821,
        protocol_version: 1,
        device_name: "Android Device",
      });
      expect(isDiscoverySupportedVersion(result.info)).toBe(true);
    }
  });

  it("trims the device display name", () => {
    const result = parseDiscoveryInfo({ ...discovery.peer, device_name: "  Android Device  " });
    expect(result).toEqual({
      ok: true,
      info: { ...discovery.peer, device_name: "Android Device" },
    });
  });

  it("ignores unknown extra TXT keys (forward-compatible metadata)", () => {
    const result = parseDiscoveryInfo(discovery.peer_with_extra_txt);
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(result.info.device_name).toBe("Android Device");
      expect(result.info.port).toBe(45821);
    }
  });

  it("rejects every malformed discovery information case from the fixture", () => {
    for (const ctx of discovery.invalid_discovery_infos) {
      const input = withMutations(discovery.peer, ctx.mutations);
      const result = parseDiscoveryInfo(input);
      expect(result, ctx.name).toEqual({ ok: false, problem: ctx.problem });
    }
  });

  it("accepts an unsupported protocol version as valid metadata but flags it", () => {
    const result = parseDiscoveryInfo(discovery.unsupported_version_peer);
    expect(result.ok).toBe(true);
    if (result.ok) {
      expect(isDiscoverySupportedVersion(result.info)).toBe(false);
    }
  });

  it("exposes only the defined local problem tags", () => {
    expect(DISCOVERY_PROBLEMS).toEqual([
      "INVALID_SERVICE",
      "INVALID_HOSTNAME",
      "INVALID_PORT",
      "INVALID_PROTOCOL_VERSION",
      "INVALID_DEVICE_NAME",
    ]);
  });

  it("registry collapses duplicate advertisements into one logical peer", () => {
    const registry = createDiscoveryRegistry();
    registry.upsert(discovery.peer);
    registry.upsert(discovery.peer);
    registry.upsert(discovery.peer_with_extra_txt);
    expect(registry.size).toBe(1);
    expect(registry.list()[0]?.device_name).toBe("Android Device");
  });

  it("identifies peers by hostname:port", () => {
    expect(deviceIdentity("receiver-device.local", 45821)).toBe("receiver-device.local:45821");
    expect(discovery.device_identity.id_format).toBe("hostname:port");
    const registry = createDiscoveryRegistry();
    registry.upsert(discovery.peer);
    registry.upsert({ ...discovery.peer, hostname: "other-device.local" });
    expect(registry.size).toBe(2);
  });

  it("removes peers on disappearance and starts empty again on reappearance", () => {
    const registry = createDiscoveryRegistry();
    registry.upsert(discovery.peer);
    registry.remove("receiver-device.local", 45821);
    expect(registry.size).toBe(0);
    registry.upsert(discovery.peer);
    expect(registry.size).toBe(1);
  });

  it("clear() shuts discovery down deterministically", () => {
    const registry = createDiscoveryRegistry();
    registry.upsert(discovery.peer);
    registry.upsert({ ...discovery.peer, hostname: "other-device.local" });
    registry.clear();
    expect(registry.size).toBe(0);
    expect(registry.list()).toEqual([]);
  });
});